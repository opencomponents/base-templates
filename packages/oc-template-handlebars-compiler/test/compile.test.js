jasmine.DEFAULT_TIMEOUT_INTERVAL = 100000;

const _ = require('lodash');
const compile = require('../lib/compile.js');
const fs = require('fs-extra');
const nodeDir = require('node-dir');
const path = require('path');

const componentPath = path.join(
  __dirname,
  '../../../mocks/handlebars-component/'
);

const componentPackage = fs.readJsonSync(
  path.join(componentPath, 'package.json')
);

const package2 = _.cloneDeep(componentPackage);
const package3 = _.cloneDeep(componentPackage);
const package4 = _.cloneDeep(componentPackage);
delete package2.oc.files.static;
package3.oc.files.static = 'assets';
delete package4.oc.files.data;

const scenarios = {
  'Happy path (production=true)': {
    componentPackage,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package'),
    production: true
  },
  'Happy path (production=false)': {
    componentPackage,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package'),
    production: false
  },
  'Should handle empty static folder': {
    componentPackage: package2,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package2'),
    production: true
  },
  'Should normalise stringified static folder': {
    componentPackage: package3,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package3'),
    production: true
  },
  'Should handle server.js-less components': {
    componentPackage: package4,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package4'),
    production: true
  }
};

const execute = (options, cb) => {
  compile(options, (err, result) => {
    if (err) {
      return fs.remove(options.publishPath, () => cb(err));
    }

    result.oc.date = '';
    result.oc.files.template.version = '';
    nodeDir.paths(options.publishPath, (err2, res2) => {
      const files = res2.files
        .map(filePath => path.relative(__dirname, filePath))
        .sort();
      fs.remove(options.publishPath, () => cb(err, { result, files }));
    });
  });
};

_.each(scenarios, (scenario, testName) => {
  test(testName, done => {
    execute(scenario, (err, { result, files }) => {
      expect(err).toBeNull();
      if (result.oc.files.dataProvider) {
        result.oc.files.dataProvider.hashKey = 'dummyData';
      }
      result.oc.files.template.hashKey = 'dummyData';
      expect(result).toMatchSnapshot();
      expect(files).toMatchSnapshot();
      done();
    });
  });
});

test('When server compilation fails should return an error', done => {
  const package5 = _.cloneDeep(componentPackage);
  package5.oc.files.data = 'not-found.js';

  const options = {
    componentPackage: package5,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package5')
  };

  execute(options, err => {
    expect(err).toContain('Entry module not found');
    done();
  });
});

test('When files writing fails should return an error', done => {
  const spy = jest
    .spyOn(fs, 'ensureDir')
    .mockImplementation(jest.fn((a, cb) => cb('sorry I failed')));

  const options = {
    componentPackage,
    ocPackage: {
      version: '1.0.0'
    },
    componentPath,
    publishPath: path.join(componentPath, '_compile-tests-package6')
  };

  execute(options, err => {
    expect(err).toMatchSnapshot();
    spy.mockRestore();
    done();
  });
});
