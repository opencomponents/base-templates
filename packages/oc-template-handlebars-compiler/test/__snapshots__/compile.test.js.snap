// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Happy path 1`] = `
Object {
  "dependencies": Object {
    "lodash": "",
  },
  "description": "",
  "name": "handlebars-component",
  "oc": Object {
    "date": "",
    "files": Object {
      "dataProvider": Object {
        "hashKey": "71a1269026ed6dcc29299ed85a8a8bd5660723c3",
        "src": "server.js",
        "type": "node.js",
      },
      "static": Array [
        "assets",
      ],
      "template": Object {
        "hashKey": "2112a7cce970d8e5eb8a60c5ada1c4d822b22b0e",
        "src": "template.js",
        "type": "handlebars",
        "version": "",
      },
    },
    "packaged": true,
    "version": "1.0.0",
  },
  "repository": "",
  "version": "1.0.0",
}
`;

exports[`Happy path 2`] = `
Array [
  "../../../mocks/handlebars-component/_compile-tests-package/assets/css/style.css",
  "../../../mocks/handlebars-component/_compile-tests-package/assets/img/test.png",
  "../../../mocks/handlebars-component/_compile-tests-package/assets/js/lib.js",
  "../../../mocks/handlebars-component/_compile-tests-package/package.json",
  "../../../mocks/handlebars-component/_compile-tests-package/server.js",
  "../../../mocks/handlebars-component/_compile-tests-package/server.js.map",
  "../../../mocks/handlebars-component/_compile-tests-package/template.js",
]
`;

exports[`Should handle empty static folder 1`] = `
Object {
  "dependencies": Object {
    "lodash": "",
  },
  "description": "",
  "name": "handlebars-component",
  "oc": Object {
    "date": "",
    "files": Object {
      "dataProvider": Object {
        "hashKey": "71a1269026ed6dcc29299ed85a8a8bd5660723c3",
        "src": "server.js",
        "type": "node.js",
      },
      "static": Array [],
      "template": Object {
        "hashKey": "2112a7cce970d8e5eb8a60c5ada1c4d822b22b0e",
        "src": "template.js",
        "type": "handlebars",
        "version": "",
      },
    },
    "packaged": true,
    "version": "1.0.0",
  },
  "repository": "",
  "version": "1.0.0",
}
`;

exports[`Should handle empty static folder 2`] = `
Array [
  "../../../mocks/handlebars-component/_compile-tests-package2/package.json",
  "../../../mocks/handlebars-component/_compile-tests-package2/server.js",
  "../../../mocks/handlebars-component/_compile-tests-package2/server.js.map",
  "../../../mocks/handlebars-component/_compile-tests-package2/template.js",
]
`;

exports[`Should handle server.js-less components 1`] = `
Object {
  "dependencies": Object {
    "lodash": "",
  },
  "description": "",
  "name": "handlebars-component",
  "oc": Object {
    "date": "",
    "files": Object {
      "static": Array [
        "assets",
      ],
      "template": Object {
        "hashKey": "2112a7cce970d8e5eb8a60c5ada1c4d822b22b0e",
        "src": "template.js",
        "type": "handlebars",
        "version": "",
      },
    },
    "packaged": true,
    "version": "1.0.0",
  },
  "repository": "",
  "version": "1.0.0",
}
`;

exports[`Should handle server.js-less components 2`] = `
Array [
  "../../../mocks/handlebars-component/_compile-tests-package4/assets/css/style.css",
  "../../../mocks/handlebars-component/_compile-tests-package4/assets/img/test.png",
  "../../../mocks/handlebars-component/_compile-tests-package4/assets/js/lib.js",
  "../../../mocks/handlebars-component/_compile-tests-package4/package.json",
  "../../../mocks/handlebars-component/_compile-tests-package4/template.js",
]
`;

exports[`Should normalise stringified static folder 1`] = `
Object {
  "dependencies": Object {
    "lodash": "",
  },
  "description": "",
  "name": "handlebars-component",
  "oc": Object {
    "date": "",
    "files": Object {
      "dataProvider": Object {
        "hashKey": "71a1269026ed6dcc29299ed85a8a8bd5660723c3",
        "src": "server.js",
        "type": "node.js",
      },
      "static": Array [
        "assets",
      ],
      "template": Object {
        "hashKey": "2112a7cce970d8e5eb8a60c5ada1c4d822b22b0e",
        "src": "template.js",
        "type": "handlebars",
        "version": "",
      },
    },
    "packaged": true,
    "version": "1.0.0",
  },
  "repository": "",
  "version": "1.0.0",
}
`;

exports[`Should normalise stringified static folder 2`] = `
Array [
  "../../../mocks/handlebars-component/_compile-tests-package3/assets/css/style.css",
  "../../../mocks/handlebars-component/_compile-tests-package3/assets/img/test.png",
  "../../../mocks/handlebars-component/_compile-tests-package3/assets/js/lib.js",
  "../../../mocks/handlebars-component/_compile-tests-package3/package.json",
  "../../../mocks/handlebars-component/_compile-tests-package3/server.js",
  "../../../mocks/handlebars-component/_compile-tests-package3/server.js.map",
  "../../../mocks/handlebars-component/_compile-tests-package3/template.js",
]
`;

exports[`When files writing fails should return an error 1`] = `"template.hbs compilation failed - sorry I failed"`;
